include/swift/SILOptimizer/Utils/PrunedLiveness.h

```diff
@@ -0,0 +1,222 @@
+//===--- PrunedLiveness.hpp - Compute liveness from selected uses ---------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+///
+/// Incrementally compute and represent basic block liveness of a single live
+/// range. The live range is defined by points in the CFG, independent of any
+/// particular SSA value. The client initializes liveness with a set of
+/// definition blocks, typically a single block. The client then incrementally
+/// updates liveness by providing a set of "interesting" uses one at a time.
+///
+/// This supports discovery of pruned liveness during control flow traversal. It
+/// is not tied to a single SSA value and allows the client to select
+/// interesting uses while ignoring other uses.
+///
+/// The PrunedLiveBlocks result maps each block to its current liveness state:
+/// Dead, LiveWithin, LiveOut.
+///
+/// A LiveWithin block has a liveness boundary within the block. The client can
+/// determine the boundary's intruction position by searching for the last use.
+///
+/// LiveOut indicates that liveness extends into a successor edges, therefore,
+/// no uses within that block can be on the liveness boundary, unless that use
+/// occurs before a def in the same block.
+///
+/// All blocks start Dead. They transition to another live state each time a new
```

Comment:
First let me say I really love this entire file level comment! I think it is adding the detail I was worried was missing in the previous iteration.

That being said, I think this section needs the following changes:

1. All blocks do not start dead. Instead, the def block is LiveWithin and /all other blocks/ are dead. This is an important detail in the initial condition of the dataflow.
2. I would add an example of an incremental liveness update. I think it would be fine if it was just a single transition, from an initial condition to adding a single use (so a single def/use pair). Something like:

```
/// Example 3. Incremental Liveness Update
///
/// Initial State:
///
///  -----
/// | Def | [LiveWithin]
///  -----
///    |
///  -----
/// | Use | [Dead]
///  -----
///
/// After adding use:
///
///  -----
/// | Def | [LiveOut]
///  -----
///    |
///  -----
/// | Use | [LiveWithin]
///  -----
```